pkg/distributor/distributor.go

```diff
@@ -41,18 +51,34 @@ var (
 	rfStats           = usagestats.NewInt("distributor_replication_factor")
 )
 
+type ShardStreamsConfig struct {
+	Enabled             bool `yaml:"enabled"`
+	DebugLoggingEnabled bool `yaml:"debug_logging_enabled"`
+}
+
 // Config for a Distributor.
 type Config struct {
 	// Distributors ring
 	DistributorRing RingConfig `yaml:"ring,omitempty"`
 
 	// For testing.
 	factory ring_client.PoolFactory `yaml:"-"`
+
+	// ShardStreams configures wether big streams should be sharded or not.
+	ShardStreams ShardStreamsConfig `yaml:"shard_streams"`
 }
 
 // RegisterFlags registers distributor-related flags.
 func (cfg *Config) RegisterFlags(fs *flag.FlagSet) {
 	cfg.DistributorRing.RegisterFlags(fs)
+	fs.BoolVar(&cfg.ShardStreams.Enabled, "distributor.stream-sharding.enabled", false, "Automatically shard streams to keep them under the per-stream rate limit")
+	fs.BoolVar(&cfg.ShardStreams.DebugLoggingEnabled, "distributor.stream-sharding.debug-logging-enabled", false, "Enable debug logging when sharding streams")
+}
+
+// StreamSharder manages the state necessary to shard streams.
+type StreamSharder interface {
+	ShardsFor(stream logproto.Stream) (int, bool)
+	IncreaseShardsFor(stream logproto.Stream)
```

Comment:
Is this called in production? I cannot find it.

Reply:
Not yet. `IncreaseShards` will be called as part of a follow-on PR that triggers stream sharding based on errors from the ingester